lib/Sema/CSSimplify.cpp

```diff
@@ -2155,8 +2155,7 @@ ConstraintSystem::matchTypes(Type type1, Type type2, ConstraintKind kind,
             PointerTypeKind type1PointerKind;
             bool type1IsPointer{
                 unwrappedType1->getAnyPointerElementType(type1PointerKind)};
-            bool optionalityMatches =
-                type1OptionalKind == OTK_None || type2OptionalKind != OTK_None;
+            bool optionalityMatches = !type1IsOptional || type2IsOptional;
```

Comment:
This is a literal translation that removes the use of OptionalTypeKind, but I am wondering if the original code was meant to be checking that they are either both Optional<T> or neither is Optional<T>, given the name of the bool the result is stored to.

@jrose-apple, it looks like you added the original code. What was the intent here?

Reply:
I think it's supposed to accept all three valid conversions (non -> non, non -> opt, opt -> opt), so maybe `optionalityIsCompatible` would have been a better name.